File: Array/C/M442-find-all-duplicates-in-an-array.c
// time:  O(n)
// space: O(1)

/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* findDuplicates(int* nums, int numsSize, int* returnSize) {
    int i;
	
    for(i = 0; i < numsSize; i++){
        nums[(nums[i] - 1 + numsSize) % numsSize] += numsSize;
    }
    
	*returnSize = 0;
    for(i = 0; i < numsSize; i++){
        if(nums[i] > 2 * numsSize){
            nums[*returnSize] = i + 1;
            (*returnSize)++;
        }
    }
    return nums;
}

// I ignored the condition that every element would appear either once or twice, so
// I just find out all elements that appear more than once. This solution is more simple.

int* findDuplicates(int * nums, int numsSize, int * returnSize){
	int i, j;
	int* answer = (int*) malloc((numsSize >> 1) * sizeof(int));
	*returnSize = 0;

	for (i = 0; i < numsSize; ++i) {
		j = abs(nums[i]) - 1;

		if (nums[j] < 0){
			answer[(*returnSize)] = j + 1;
			(*returnSize)++;
		} else {
			nums[j] = -nums[j];
		}
	}
	return answer;
}

File: Array/Python3/M442-find-all-duplicates-in-an-array.py
# time:  O(n)
# space: O(1)

class Solution:
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        res = []
        for x in nums:
            if nums[abs(x) - 1] < 0:
                res.append(abs(x))
            else:
                nums[abs(x) - 1] *= -1
                
        return res

    # a faster solution
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        l=[0]*(len(nums)+1)
        res=[]
        for i in nums:
            if l[i]==1:
                res.append(i)
            else:
                l[i]=1
        return res
